### PR TITLE
docs: non-checkpoint-branch workflow

### PR DESCRIPTION
**Pull request by Jinjiang**
An update to follow this workflow demo:
https://github.com/Jinjiang/translation-workflow-demo

This is considered an alternative approach to #1824 

Changes:
1. Included the changes in #1824 
2. Added a new npm script to generate a file named `translation-status.json` to record translation checkpoints. This script should be run when each time we sync up translations for any language.
3. Simplified the workflow by removing the maintenance of checkpoint branches.

Thanks.

### CODE REVIEW

**Comment by Jinjiang** [comment]
@posva This is ready for review. Thanks.

**Comment by posva** [comment]
@Jinjiang Thanks! Really sorry for the delay, I will give it a look in 1-2 weeks

**Comment by nazarepiedady** [comment]
@jinjiang I really liked the idea, I hope it goes ahead.

File: .github/contributing.md [hunk 1]
```diff
@@ -117,40 +117,50 @@ Currently, all the docs can be found in `packages/docs`. It contains the English
...
 
-If you want to start translating the docs in a new language:
+If you want to start translating the docs in a _new_ language:
 
 1. Create the corresponding `<lang>` sub-folder for your translation.
 2. Modify the i18n config in `.vitepress` sub-folder.
 3. Translate the docs and run the doc site to self-test locally.
-4. Once you have done all above, create a pull request to our GitHub repo.
+4. Create a checkpoint for your language by running `pnpm run docs:translation-status <lang> [<commit>]`. A checkpoint is the hash and date of the latest commit when you do the translation. The checkpoint information would be stored in the status file `packages/docs/.vitepress/translation-status.json`. _It's important for the long-term maintenance since all the further translation sync-ups would be based on their previous checkpoints._ Usually you can skip the commit argument because the default value is `main`.
+5. Commit all the changes and create a pull request to our GitHub repo. Then after the approval and merge by the maintainers. The translation would be affected.
```
Comment (posva) [inline .github/contributing.md:130]: Unless I'm misunderstanding, this doesn't add any information, I think it's simpler without it

```suggestion
5. Commit all the changes and create a pull request to our GitHub repo.
```

File: .github/contributing.md [hunk 2]
```diff
@@ -117,40 +117,50 @@ Currently, all the docs can be found in `packages/docs`. It contains the English
...
 
-If you want to maintain a existing translation:
+To be noticed, on the top of each translation page, there would be a paragraph to indicate their translation status. So the users can know whether the translation is up-to-date or not.
```
Comment (posva) [inline .github/contributing.md:132]: Let's make it clearer:

```suggestion
We will have a paragraph right at the top of each translation page that shows the status of the translation. That way, users can easily figure out if the translation is up-to-date or lags behind the English version.
```

File: .github/contributing.md [hunk 3]
```diff
@@ -117,40 +117,50 @@ Currently, all the docs can be found in `packages/docs`. It contains the English
...
+If you want to maintain an existing translation:
 
-For more real examples, please check out [all the PRs with title "docs(zh): sync" after 2023-01-01](https://github.com/vuejs/router/pulls?q=is%3Apr+created%3A%3E2023-01-01+docs%28zh%29+sync).
+1. See what translation you need to do to sync up with the original docs. There are 2 popular ways:
+    1. Via the GitHub Compare page: https://github.com/vuejs/router/compare/ (only see the changes in `packages/docs/*`) from the checkpoint hash to `main` branch. You can find the checkpoint hash for your language via the translation status file `packages/docs/.vitepress/translation-status.json`.
+   2. Via a local command: `pnpm run docs:compare-to-translate <lang> [<commit>]`.
+2. Create your own branch and start the translation update, following the previous comparison.
+3. Same to step 4 and step 5 in the “starting a new language” workflow. Once you have done all above, create a pull request and then wait for the maintainers to approve and merge it.
```
Comment (posva) [inline .github/contributing.md:140]: There are a few issues with this section:

- no introductory sentence. It's as if the 1st point was that but I think it's missing something along the lines of Keeping translations updated
- inconsistent spacing (3 and 4, should be 2 spaces)
- The sample URL should be a real example so users can actually replace things in that URL to test it out

Comment (posva) [inline .github/contributing.md:140]: I think it's better to explicitly add the steps 4 and 5 with adapted commands

File: .github/contributing.md [hunk 4]
```diff
@@ -117,40 +117,50 @@ Currently, all the docs can be found in `packages/docs`. It contains the English
...
 
-1. Ensure you create a _checkpoint branch_ (for example, a branch named sync). This branch should always align with the commit of the original documentation that corresponds to your most recent translation.
-2. Utilize the diff result between the latest official repository and your own by using the git diff command or the GitHub Compare page to guide your translation.
-3. Complete the translation process.
-4. Update the _checkpoint branch_ accordingly.
+- Ensure you maintain the _checkpoint_ properly. And also ensure the _translation status_ is well-displayed on the top of each translation page.
+- Utilize the diff result between the latest official repository and your own checkpoint to guide your translation.
+
+Tips: you can add the official repo as a remote to your forked repo, and then still able to run `pnpm run docs:translation-status <lang> [<commit>]` and `npm run docs:compare-to-translate <lang> [<commit>]` to get the checkpoint and diff result:
```
Comment (posva) [inline .github/contributing.md:151]:
```suggestion
Tip: you can add the official repo as a remote to your forked repo, this way you can still run `pnpm run docs:translation-status <lang> [<commit>]` and `npm run docs:compare-to-translate <lang> [<commit>]` to get the checkpoint and diff result:
```

File: packages/docs/compare-to-translate.mjs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+// The path to the translation status file.
+const STATUS_FILE_PATH = './.vitepress/translation-status.json'
+
+const usage = `
+Usage: pnpm run docs:compare-to-translate <locale> [<commit>]
+  locale: The target locale to compare.
+  commit: The target commit to compare. It could be a branch, a tag, or a hash. Default to 'main'.`
+
+const getLocaleHash = async (lang) => {
```
Comment (posva) [inline packages/docs/compare-to-translate.mjs:13]:
```suggestion
async function getLocaleHash (lang) {
```

File: packages/docs/compare-to-translate.mjs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    const content = await readFile(STATUS_FILE_PATH, 'utf8')
+    const data = JSON.parse(content)
+    return data[lang]?.hash
+  } catch (err) {
+    console.log('No previous status file. Will create a new one.')
+  }
+}
+
+async function main() {
+  if (!process.argv[2] || process.argv[2] === '--help' || process.argv[2] === '-h') {
```
Comment (posva) [inline packages/docs/compare-to-translate.mjs:24]: Probably safer:

```suggestion
  if (process.argv.find(arg => arg === '--help' || arg === '-h')) {
```

File: packages/docs/compare-to-translate.mjs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    console.log(usage)
+    return
+  }
+
+  const locale = process.argv[2]
+  const commit = process.argv[3] || 'main'
+
+  const hash = await getLocaleHash(locale)
+  if (hash) {
+    console.log(`The last checkpoint of docs(${locale}) is ${hash}.\n`)
```
Comment (posva) [inline packages/docs/compare-to-translate.mjs:34]: for more readability:

```suggestion
    console.log(`The last checkpoint of docs(${locale}) is "${hash}".\n`)
```

File: packages/docs/compare-to-translate.mjs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    const git = simpleGit()
+    const result = await git.diff([`${hash}..${commit}`, '.'])
+    console.log(result)
+    console.log(`\nAfter finishing the translation, you can run "pnpm run docs:translation-status ${locale} ${hash}" or "pnpm run docs:translation-status ${locale}${commit !== 'main' ? ' ' + commit : ''}" to update the translation status file.`)
```
Comment (posva) [inline packages/docs/compare-to-translate.mjs:38]: For easy copy-paste

```suggestion
    console.log(`\nAfter finishing the translation, you can run\n"pnpm run docs:translation-status ${locale} ${hash}"\nor\n"pnpm run docs:translation-status ${locale}${commit !== 'main' ? ' ' + commit : ''}"\nto update the translation status file.`)
```

File: packages/docs/generate-translation-status.mjs [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+// The path to the translation status file.
+const STATUS_FILE_PATH = './.vitepress/translation-status.json'
+
+const usage = `
+Usage: pnpm run docs:translation-status <locale> [<commit>]
+  locale: The target locale to update.
+  commit: The target commit to update. It could be a branch, a tag, or a hash. Default to 'main'.`
+
+const getCommitInfo = async (commit) => {
```
Comment (posva) [inline packages/docs/generate-translation-status.mjs:13]:
```suggestion
async function getCommitInfo (commit) {
```

File: packages/docs/generate-translation-status.mjs [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    const git = simpleGit()
+    const log = await git.log([commit, '-n', '1'])
+    const { hash, date } = log.latest
+    return { hash, date: new Date(date).toISOString().substring(0, 10) }
+  } catch (err) {
+    return { hash: '', date: '' }
+  }
+}
+
+const writeLangMap = async (lang, hash, date) => {
```
Comment (posva) [inline packages/docs/generate-translation-status.mjs:24]:
```suggestion
async function writeLangMap (lang, hash, date) {
```

File: packages/docs/generate-translation-status.mjs [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+  }
+  await writeFile(STATUS_FILE_PATH, JSON.stringify(data, null, 2))
+}
+
+async function main() {
+  if (
+    process.argv.length < 3 ||
+    process.argv[2] === '--help' ||
+    process.argv[2] === '-h'
+  ) {
```
Comment (posva) [inline packages/docs/generate-translation-status.mjs:46]:
```suggestion
  if (process.argv.find(arg => arg === '--help' || arg === '-h')) {
```

File: packages/docs/generate-translation-status.mjs [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    console.log(usage)
+    return
+  }
+
+  const locale = process.argv[2]
+  const commit = process.argv[3] || 'main'
+
+  const { hash, date } = await getCommitInfo(commit)
+  if (!hash) {
+    console.log(`❌ No commit found for ${commit}.`)
```
Comment (posva) [inline packages/docs/generate-translation-status.mjs:56]:
```suggestion
    console.log(`❌ No commit found for "${commit}".`)
```

File: packages/docs/generate-translation-status.mjs [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    return
+  } else {
+    await writeLangMap(locale, hash, date)
+    console.log(`✅ Updated ${locale} to ${hash} (${date})`)
```
Comment (posva) [inline packages/docs/generate-translation-status.mjs:60]:
```suggestion
    console.log(`✅ Updated ${locale} to "${hash}" (${date})`)
```

**Review by posva — CHANGES_REQUESTED** [review]
Thanks a lot @Jinjiang and sorry again for the delay 🙏

File: .github/contributing.md (hunk 3 above)
Comment (Jinjiang) [reply .github/contributing.md]: @posva for spacing, I found 2 spaces in front of ordered list won't be recognized as indents in the Markdown preview. So I kept 4 spaces instead. The 3 spaces was a mistake, sorry about that (however they can also be recognized as indents).

For the rest part I modified a little accordingly. Please take a further look. Thanks.

**Comment by Jinjiang** [comment]
@posva hey, do you think we can merge this or there is anything further to improve? :-)

**Review by posva — APPROVED** [review]

**Comment by Jinjiang** [comment]
@nazarepiedady now this workflow has been officially confirmed. feel free to translate the vue router docs in this way. 😉

I will also write something about this in the vuejs-translations org later.

Thanks.

**Comment by nazarepiedady** [comment]
@Jinjiang, I will start making the necessary changes to organize the work I am doing over the Portuguese translation to match the current conclusions about how the translation workflow should be.

**Pull request merged**